lib/node_modules/@stdlib/math/base/special/cosc/test/test.js

```diff
@@ -0,0 +1,196 @@
+var NINF = require( '@stdlib/constants/math/float64-ninf' );
+var EPS = require( '@stdlib/constants/math/float64-eps' );
+var abs = require( '@stdlib/math/base/special/abs' );
+var cosc = require( './../lib' );
+
+
+// FIXTURES //
+
+var data = require( './fixtures/julia/data.json' );
+var largeNegative = require( './fixtures/julia/large_negative.json' );
+var largePositive = require( './fixtures/julia/large_positive.json' );
+var tinyNegative = require( './fixtures/julia/tiny_negative.json' );
+var tinyPositive = require( './fixtures/julia/tiny_positive.json' );
+
+
+// TESTS //
+
+tape( 'main export is a function', function test( t ) {
+	t.ok( true, __filename );
+	t.equal( typeof cosc, 'function', 'main export is a function' );
+	t.end();
+});
+
+tape( 'the function computes the derivative of the cardinal sine', function test( t ) {
+	var expected;
+	var delta;
+	var tol;
+	var x;
+	var y;
+	var i;
+
+	x = data.x;
+	expected = data.expected;
+
+	for ( i = 0; i < x.length; i++ ) {
+		y = cosc( x[i] );
+		if ( y === expected[ i ] ) {
+			t.equal( y, expected[ i ], 'x: '+x[i]+'. Expected: '+expected[i] );
+		} else {
+			delta = abs( y - expected[i] );
+			tol = 2.0 * EPS * abs( expected[i] );
+			t.ok( delta <= tol, 'within tolerance. x: '+x[i]+'. Value: '+y+'. Expected: '+expected[i]+'. Tolerance: '+tol+'.' );
+		}
+	}
+	t.end();
+});
+
+tape( 'the function computes the derivative of the cardinal sine (large negative)', function test( t ) {
+	var expected;
+	var delta;
+	var tol;
+	var x;
+	var y;
+	var i;
+
+	x = largeNegative.x;
+	expected = largeNegative.expected;
+
+	for ( i = 0; i < x.length; i++ ) {
+		y = cosc( x[i] );
+		if ( y === expected[ i ] ) {
+			t.equal( y, expected[ i ], 'x: '+x[i]+'. Expected: '+expected[i] );
+		} else {
+			delta = abs( y - expected[i] );
+			tol = 2.0 * EPS * abs( expected[i] );
+			t.ok( delta <= tol, 'within tolerance. x: '+x[i]+'. Value: '+y+'. Expected: '+expected[i]+'. Tolerance: '+tol+'.' );
+		}
+	}
+	t.end();
+});
+
+tape( 'the function computes the derivative of the cardinal sine (large positive)', function test( t ) {
+	var expected;
+	var delta;
+	var tol;
+	var x;
+	var y;
+	var i;
+
+	x = largePositive.x;
+	expected = largePositive.expected;
+
+	for ( i = 0; i < x.length; i++ ) {
+		y = cosc( x[i] );
+		if ( y === expected[ i ] ) {
+			t.equal( y, expected[ i ], 'x: '+x[i]+'. Expected: '+expected[i] );
+		} else {
+			delta = abs( y - expected[i] );
+			tol = 2.0 * EPS * abs( expected[i] );
+			t.ok( delta <= tol, 'within tolerance. x: '+x[i]+'. Value: '+y+'. Expected: '+expected[i]+'. Tolerance: '+tol+'.' );
+		}
+	}
+	t.end();
+});
+
+tape( 'the function computes the derivative of the cardinal sine (tiny negative)', function test( t ) {
+	var expected;
+	var delta;
+	var tol;
+	var x;
+	var y;
+	var i;
+
+	x = tinyNegative.x;
+	expected = tinyNegative.expected;
+
+	for ( i = 0; i < x.length; i++ ) {
+		if (expected[i] > 1e+100) // Jump over Julia fixture values that are out of wack
```

Comment:
How large is the difference? We may need to investigate the source of the discrepancy. Given that the underlying trig functions which are called in this implementation have reasonable correspondence to the Julia implementations, I am a bit surprised that values are significantly different.

Reply:
If I run the Julia cosc function on the following two values of x I get:
```
x = -1.0e-200
cosc(x) returns 0.0

x= -9.980079681474104e-201
cosc(x) returns 2.2248780772483203e184
```
These abnomal values only occur for every tenth value of x in the in the tiny negative and tiny postive range.

Reply:
Interesting! Okay, I tracked it down. Apparently, this is due to a small precision error in the following operation in Julia:

```julia
> -9.980079681474104e-201 * pi
-3.1353345009559805e-200
> sinpi(-9.980079681474104e-201)
-3.135334500955981e-200
```

That very small difference means that, when the latter is divided by the former, the result is not exactly `1.0`, but `1.0000000000000002`. Accordingly,

```julia
> x = -9.980079681474104e-201;
> cospi(x) - (sinpi(x)/(pi*x))
-2.220446049250313e-16
```
which is double-precision floating-point epsilon. When divided by `-9.980079681474104e-201`, this gives a very large number.

In contrast, in V8 (Node.js),

```javascript
> PI*-9.980079681474104e-201
-3.1353345009559805e-200
> base.sinpi(-9.980079681474104e-201)
-3.1353345009559805e-200
```
i.e., the return values are the same (although, this **may** vary across JavaScript engines). And hence, we get the return value of `0.0`.

In short, it seems the tests happened to, in a manner of speaking, stumble on a change in precision that occurs within Julia's [`sinpi`](https://github.com/JuliaLang/julia/blob/dee7d77c199c4776cf4e06afb7ce920f09ad3c8c/base/special/trig.jl#L764) implementation, which does **not** occur in our [`sinpi`](https://github.com/stdlib-js/stdlib/blob/5ec19d2b8a855c2a6e4809092a87140127a6ae2a/lib/node_modules/%40stdlib/math/base/special/sinpi/lib/sinpi.js) implementation.

While interesting from a numerical POV, I am not sure how much we need to dwell on this in the tests. We could, e.g., instead of generating linearly spaced values of `x`, just generate a bunch of random tiny values. Provided these values don't fall prey to the same issue, we should be more or less in agreement with Julia.

@Planeshifter You have any thoughts on this?

Reply:
We could switch out the existing tiny negative fixtures by a bunch of tiny values for which the error doesn't occur in Julia. Might be worthwhile to add a comment somewhere to keep a record of it. But seems everything is in the clear for us here.

Reply:
@Rinbo You mind updating the test fixtures to this end?

Reply:
@kgryte I updated the fixtures with ten examples each for tiny negative and tiny postitive that passes the test. But perhaps that is not enough? All in all there are 3032 tests though.